src/ir/auth_logic/lowering_ast_datalog.cc

```diff
@@ -262,11 +273,129 @@ std::vector<datalog::Rule> LoweringToDatalogPass::QueriesToDLIR(
   return utils::MapIter<datalog::Rule>(queries, [this](const Query& query) {
     auto [main_pred, not_used] = FactToDLIR(query.principal(), query.fact());
     main_pred = PushPrincipal("says_", query.principal(), main_pred);
-    datalog::Predicate lhs(query.name(), {"dummy_var"}, datalog::kPositive);
+    datalog::Predicate lhs(query.name(), {R"("dummy_var")"},
+                           datalog::kPositive);
     return datalog::Rule(lhs, {main_pred, kDummyPredicate});
   });
 }
 
+std::vector<datalog::RelationDeclaration>
+LoweringToDatalogPass::RelationDeclarationToDLIR(
+    const std::vector<datalog::RelationDeclaration>& relation_declarations) {
+  // Note that the additional relation declarations that are generated for
+  // (possibly nested) cansay expressions are based on an environment
+  // that tracks the maximum depth with which they appear. This environment is
+  // populated by side-effect while translating the program body. As a result,
+  // it is assumed that this function is called after the body is translated.
+
+  // The base relation declarations are the same as the ones from the
+  // surface program plus another one for "canActAs"
+  std::vector<datalog::RelationDeclaration> base_declarations =
+      relation_declarations;
+  base_declarations.push_back(datalog::RelationDeclaration(
+      "canActAs", false,
+      {datalog::Argument(
+           "p1", datalog::ArgumentType(datalog::ArgumentType::Kind::kPrincipal,
+                                       "Principal")),
+       datalog::Argument(
+           "p2", datalog::ArgumentType(datalog::ArgumentType::Kind::kPrincipal,
+                                       "Principal"))}));
```

Comment:
Can't you add this directly to the transformed_declarations afterwards? That way you don't have to make a copy of `relation_declarations` at line 293 above?

Reply:
You are right. That would be efficient. I will do that.